processor/src/test/java/com/linecorp/decaton/processor/runtime/ProcessPipelineTest.java

```diff
@@ -89,65 +98,134 @@ private static TaskRequest taskRequest() {
+        pipeline.scheduleThenProcess(request);
+        verify(schedulerMock, never()).schedule(any());
+        verify(processorMock, never()).process(any(), any());
+        verify(request.completion(), times(1)).complete();
     }
 
     @Test
     public void testExtract_PurgeRawRequestBytes() {
         when(extractorMock.extract(any()))
                 .thenReturn(new DecatonTask<>(TaskMetadata.fromProto(REQUEST.getMetadata()), TASK, TASK.toByteArray()));
 
-        TaskRequest requestSpy = spy(taskRequest());
-
-        pipeline.extract(requestSpy);
-
-        verify(requestSpy, times(1)).purgeRawRequestBytes();
-    }
+        TaskRequest request = taskRequest();
+        pipeline.extract(request);
 
-    static class ProcessException extends RuntimeException {
+        assertNull(request.rawRequestBytes());
     }
 
-    @Test(expected = ProcessException.class)
-    public void testProcess_SynchronousFailure() throws InterruptedException {
+    @Test
+    public void testScheduleThenProcess_SynchronousFailure() throws InterruptedException {
         DecatonTask<HelloTask> task = new DecatonTask<>(TaskMetadata.fromProto(REQUEST.getMetadata()), TASK, TASK.toByteArray());
+        when(extractorMock.extract(any())).thenReturn(task);
 
-        when(extractorMock.extract(any()))
-                .thenReturn(task);
+        doThrow(new RuntimeException()).when(processorMock).process(any(), eq(TASK));
 
-        doThrow(new ProcessException()).when(processorMock).process(any(), eq(TASK));
+        TaskRequest request = taskRequest();
+        // Checking exception doesn't bubble up
+        pipeline.scheduleThenProcess(request);
+        verify(request.completion(), times(1)).complete();
+    }
 
-        pipeline.process(taskRequest(), task);
+    @Test(timeout = 5000)
+    public void testScheduleThenProcess_Terminate() throws InterruptedException {
+        DecatonTask<HelloTask> task = new DecatonTask<>(TaskMetadata.fromProto(REQUEST.getMetadata()), TASK, TASK.toByteArray());
+        when(extractorMock.extract(any())).thenReturn(task);
+
+        CountDownLatch atSchedule = new CountDownLatch(1);
+        CountDownLatch closeLatch = new CountDownLatch(1);
+        doAnswer(invocation -> {
+            atSchedule.countDown();
+            closeLatch.await();
+            return null;
+        }).when(schedulerMock).schedule(any());
+
+        ExecutorService executor = Executors.newFixedThreadPool(1);
+        executor.execute(() -> {
+            try {
+                pipeline.scheduleThenProcess(taskRequest());
+            } catch (InterruptedException e) {
+                fail("Fail by exception: " + e);
+            }
+        });
+        atSchedule.await();
+        pipeline.close();
+        closeLatch.countDown();
+
+        executor.shutdown();
+        // Checking it actually returns
+        executor.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
+        verify(pipeline, never()).process(any(), any());
```

Comment:
ProcessPipeline is now responsible for managing offset completion, so better to check competion.compete() is not called ?

Reply:
good point, let me add that.